cpp/src/reductions/reductions.cu

```diff
@@ -39,234 +35,160 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
-        int gridsize = (output_size < REDUCTION_BLOCK_SIZE?
-                        output_size : REDUCTION_BLOCK_SIZE);
-
-        // launch kernel
-        gpu_reduction_op<<<gridsize, blocksize>>>(
-            // inputs
-            data, valid, size,
-            // output
-            output,
-            // action
-            functor,
-            // identity
-            identity,
-            // loader
-            loader
-        );
+    // First thread of each block stores the result.
+    if (tid == 0){
+        cudf::genericAtomicOperation(&results[0], agg, functor);
     }
+}
 
-};
-
+template<typename T_in, typename T_out, typename Op>
+void ReduceOp(const gdf_column *input,
+                   gdf_scalar* scalar, cudaStream_t stream)
+{
+    T_out identity = Op::Op::template identity<T_out>();
 
-struct DeviceSum {
-    typedef IdentityLoader Loader;
-    typedef DeviceSum second;
+    // allocate temporary memory for the result
+    void *result = NULL;
+    RMM_TRY(RMM_ALLOC(&result, sizeof(T_out), stream));
 
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs + rhs;
-    }
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(result, &identity,
+            sizeof(T_out), cudaMemcpyHostToDevice, stream));
+    CHECK_STREAM(stream);
 
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
-};
+    int blocksize = reduction_block_size;
+    int gridsize = (input->size + reduction_block_size -1 )
+        /reduction_block_size;
 
-struct DeviceProduct {
-    typedef IdentityLoader Loader;
-    typedef DeviceProduct second;
+    // kernel call
+    gpu_reduction_op<<<gridsize, blocksize, 0, stream>>>(
+        static_cast<const T_in*>(input->data), input->valid, input->size,
+        static_cast<T_out*>(result),
+        typename Op::Op{}, identity, typename Op::Loader{});
+    CHECK_STREAM(stream);
 
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs * rhs;
-    }
+    // read back the result to host memory
+    CUDA_TRY(cudaMemcpyAsync(&scalar->data, result,
+            sizeof(T_out), cudaMemcpyDeviceToHost, stream));
+    CHECK_STREAM(stream);
 
-    template<typename T>
-    static constexpr T identity() { return T{1}; }
-};
+    // cleanup temporary memory
+    RMM_TRY(RMM_FREE(result, stream));
 
-struct DeviceSumOfSquares {
-    struct Loader {
-        template<typename T>
-        __device__
-        T operator() (const T* ptr, int pos) const {
-            T val = ptr[pos];   // load
-            return val * val;   // squared
-        }
-    };
-    // round 2 just uses the basic sum reduction
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) const {
-        return lhs + rhs;
-    }
+    // sync the stream
+    CUDA_TRY(cudaStreamSynchronize(stream));
```

Comment:
Good point. Sorry I didn't catch this -- I saw the `stream` variable and assumed he was creating the stream. My bad. But if you remove the sync, should probably change cudaMemcpyAsync back to a synchronous memcpy since the result may be needed on the host immediately...

Reply:
@jrhemstad @harrism 
So, for the conclusion, remove `cudaStreamSynchronize` and change `cudaMemcpyAsync` into `cudaMemcpy`. Correct?